### ISSUE TITLE
Inappropriate type error with docstrings

### ISSUE THREAD

**Issue by metov** [microsoft/pylance-release#3612]
The code below is obviously valid:
```
"""Hello"""

def echo(s: str):
    print(s)

echo(__doc__)
```

And running it will print `Hello`. However, Pylance issues an error:
```
__doc__: str | None
Argument of type "str | None" cannot be assigned to parameter "s" of type "str" in function "echo"
  Type "str | None" cannot be assigned to type "str"
    Type "None" cannot be assigned to type "str" Pylance(reportGeneralTypeIssues)
No quick fixes available
```

It is clear what's going on: Evidently `__doc__` can be `None`, if no docstring was entered, and the `echo` method advertises no support for `None` inputs. However, in this case, there obviously is a docstring. Therefore, the type error is inappropriate.

**Comment by erictraut** [comment]
Yes, the static type for a module-level `__doc__` is `str | None`.

If you want to avoid the type violation error, you can add an assert. The static type analyzer will then narrow the type to `str`.

```python
assert __doc__ is not None
```

We could add logic to the type checker that mirrors the internal runtime details and modifies the static type based on the presence of a module-level doc string, but that would involves some pretty major hacks. We generally try to avoid adding such hacks to the static type checker because it leads to unintended side effects and an increased maintenance burden.

We could consider doing this in the future if there's sufficient demand. I recommend moving this to a discussion topic to see if we get signal from other pylance users that this is important. I think this is the first time in 3+ years that anyone has run into this particular issue, so I'm guessing it's pretty uncommon.

**Comment by metov** [comment]
>We could add logic to the type checker that mirrors the internal runtime details and modifies the static type based on the presence of a module-level doc string, but that would involves some pretty major hacks.

But, doesn't Pylance already parse docstrings so that it can show tooltips? If so, the logic for deciding whether a docstring is defined or not is already there. The type checker just needs to consult it.

>I'm guessing it's pretty uncommon.

That sounds unlikely to me. For example, I actually encountered this when using docopt. [Docopt](https://github.com/docopt/docopt) is a significant Python package with [almost half a million daily downloads](https://pypistats.org/packages/docopt). Yet the basic example from the docopt readme runs into this issue.

Perhaps it is more accurate to say that it has not been reported often.

I certainly see your point in that this might not be worth it if it requires a lot of special hacks. Although, I believe PyCharm doesn't have this issue, so I wonder how they avoid it.

**Comment by erictraut** [comment]
OK, I've found a way to implement that isn't too hacky. This will be included in the next release of pyright and the next pre-release version of pylance.

**Comment by metov** [comment]
Yay! Thank you very much, Eric!

**Comment by rchiodo** [comment]
There may have been some git problems. I can't find the fix in the repository. The commit is referenced here, but it's not in the main branch.

**Comment by rchiodo** [comment]
So not going to be fixed in the next release at the moment.

**Comment by erictraut** [comment]
I'm not sure what happened here. The commit seems to have been lost somehow. I've checked in the fix again.

**Comment by rchiodo** [comment]
This issue has been fixed in prerelease version 2022.11.41, which we've just released. You can find the changelog here: [CHANGELOG.md](https://github.com/microsoft/pylance-release/blob/main/CHANGELOG.md#20221141-30-November-2022-PreRelease)
